sugondat-chain/node/src/proposer.rs

```diff
@@ -0,0 +1,89 @@
+//! Wrapper around a proposer which only authors blocks when certain conditions are met.
+//!
+//! These conditions are:
+//! 1. There is at least one transaction ready to post. This is used to determine that there were
+//!    non-inherent extrinsics and avoid authoring empty blocks.
+//! 2. There is an incoming downward message from the relay chain.
+//! 3. There is a go-ahead signal for a parachain code upgrade.
+//!
+//! If any of these conditions are met, then the block is authored.
+
+use anyhow::anyhow;
+
+use sc_transaction_pool_api::TransactionPool;
+use sp_api::StorageProof;
+use sp_consensus::Proposal;
+use sp_inherents::InherentData;
+use sp_runtime::generic::Digest;
+
+use cumulus_client_consensus_proposer::{Error as ProposerError, ProposerInterface};
+use cumulus_pallet_parachain_system::relay_state_snapshot::RelayChainStateProof;
```

Comment:
Sounds like we might need to think to extract this. (Well, I guess by we I mean cumulus devs)

Reply:
Extract in what sense?

Reply:
oh, sorry, I meant that `RelayChainStateProof` seems like a generally useful primitive and as such probably would be better placed in a primitives crate. That crate would be dependent upon by places such as this one and the `cumulus_pallet_parachain_system`.